File: docs/DSA/neetcode150/src/go/sliding_window_maximum.go
package main

func maxSlidingWindow(nums []int, k int) []int {
	var deque []int
	var result []int
	l := 0
	for r, num := range nums {
		for len(deque) > 0 && nums[deque[len(deque)-1]] < num {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, r)
		for len(deque) > 0 && deque[0] < l {
			deque = deque[1:]
		}
		if r-l+1 == k {
			result = append(result, nums[deque[0]])
			l++
		}
	}
	return result
}

File: DSA/neetcode150/src/py/sliding_window_maximum.py
from typing import List
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque()
        result = []
        l = 0
        for r, num in enumerate(nums):
            while q and nums[q[-1]] < num:
                _ = q.pop()

            q.append(r)
            while q and q[0] < l:
                _ = q.popleft()

            if r-l+1 == k:
                result.append(nums[q[0]])
                l+=1
        return result